lib/phlex/csv.rb

```diff
@@ -100,22 +120,99 @@ def trim_whitespace?
 
 	# Override and set to `false` to disable CSV injection escapes or `true` to enable.
 	def escape_csv_injection?
-		nil
+		UNDEFINED
 	end
 
-	def __escape__(value)
-		value = trim_whitespace? ? value.to_s.strip : value.to_s
-		first_char = value[0]
-		last_char = value[-1]
-
-		if escape_csv_injection? && FORMULA_PREFIXES.include?(first_char)
-			# Prefix a single quote to prevent Excel, Google Docs, etc. from interpreting the value as a formula.
-			# See https://owasp.org/www-community/attacks/CSV_Injection
-			%("'#{value.gsub('"', '""')}")
-		elsif (!trim_whitespace? && (SPACE_CHARACTERS.include?(first_char) || SPACE_CHARACTERS.include?(last_char))) || value.include?('"') || value.include?(",") || value.include?("\n")
-			%("#{value.gsub('"', '""')}")
-		else
+	def __escape__(buffer, value, escape_csv_injection:, strip_whitespace:)
+		value = case value
+		when String
 			value
+		when Symbol
+			value.name
+		else
+			value.to_s
+		end
+
+		if strip_whitespace
+			value = value.strip
+
+			if escape_csv_injection
+				if FORMULA_PREFIXES_MAP[value.getbyte(0)]
+					value.gsub!('"', '""')
+					buffer << '"\'' << value << '"'
+				elsif value.match?(/[\n",]/)
+					value.gsub!('"', '""')
+					buffer << '"' << value << '"'
+				else
+					buffer << value
+				end
+			else # not escaping CSV injection
+				buffer << value
+			end
+		else # not stripping whitespace
+			if escape_csv_injection
+				first_byte = value.getbyte(0)
+
+				if FORMULA_PREFIXES_MAP[first_byte]
+					buffer << '"\'' << value.gsub('"', '""') << '"'
+				elsif value.match?(/^\s|\s$|[\n",]/)
+					buffer << '"' << value.gsub('"', '""') << '"'
+				else
+					buffer << value
+				end
+			else # not escaping CSV injection
+				if value.match?(/^\s|\s$|[\n",]/)
```

Comment:
These regexes should include the `delimiter`. We’ll have to generate them at the top and pass them down.